### PR TITLE
feat: update certificate spammer

### PR DESCRIPTION
**Pull request by atanmarko**
# Description

Significantly updated topos certificate spammer:

Fixes TP-543

## Breaking Changes

Certificate spammer is significantly refactored:

- Now it resembles more behavior of the `topos sequencer`
- Producing chain of certificates with one `source_subnet_id` derived from parameter `--signing-key` with `--interval` between certificates
- It sends certificates to one or more `--target-nodes`. Target nodes are specified as comma separated list of TCE endpoints, e.g. `--target-nodes=http://[::1]:1340,http://[::1]:1341` or `TOPOS_NETWORK_SPAMMER_TARGET_NODES=http://[::1]:1340,http://[::1]:1341`
- If `--nb-certificates` is specified it will stop after specified N certificates
- It will generate instrumentation event `NewTestCertificate` if `--otlp-agent` and `--otlp-service-name` are specified

Example:
```
topos network spam --target-nodes=http://[::1]:1340,http://[::1]:1341 --target-subnets=0x3bc19e36ff1673910575b6727a974a9abd80c9a875d41ab3e2648dbfb9e4b518,0xa00d60b2b408c2a14c5d70cdd2c205db8985ef737a7e55ad20ea32cc9e7c417c --otlp-service-name=test-debug --otlp-agent http://127.0.0.1:4317 --interval 1000
```
Connect to 2 TCE nodes `http://[::1]:1340` and `http://[::1]:1341` and every 1 second send new certificate that will have 2 specified target subnet ids `0x3bc19e36ff1673910575b6727a974a9abd80c9a875d41ab3e2648dbfb9e4b518`, `0xa00d60b2b408c2a14c5d70cdd2c205db8985ef737a7e55ad20ea32cc9e7c417c`. Source subnet id will be `0x4da962234349063b1b556150fb8b6b2a551c2e6fe74b3bebecb732f56aaf5a6e`, derived from default specified test signing private key `0x146a1bf72a18a954059f96c79fe645a099e0d999b614e2b300d08ee674f37f41`.

## PR Checklist:

- [x] I have performed a self-review of my code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [ ] I have added or updated tests that comprehensively prove my change is effective or that my feature works

### CODE REVIEW

**Comment by atanmarko** [comment]
CC @sebastiendan @foo0x29a to be aware of the changes in `topos-certificate-spammer`

File: crates/topos/src/components/network/mod.rs [hunk 1]
```diff
@@ -32,7 +33,10 @@ pub(crate) async fn handle_command(
 
             spawn(async move {
                 if let Err(error) = topos_certificate_spammer::run(config).await {
-                    panic!("Unable to execute network spam command due to : {error:?}");
+                    error!("Unable to execute network spam command due to: {error}");
+                    std::process::exit(-1);
```
Comment (Freyskeyd) [inline crates/topos/src/components/network/mod.rs:37]:
```suggestion
                    std::process::exit(1);
```

File: crates/topos/src/components/network/commands/spam.rs [hunk 2]
```diff
@@ -2,35 +2,68 @@ use clap::Args;
...
+    )]
+    pub local_key_seed: u64,
+    /// Certificates generated in one batch. Batch is generated every `batch-time-interval` milliseconds.
+    #[arg(
+        long,
+        env = "TOPOS_NETWORK_SPAMMER_CERT_PER_BATCH",
+        default_value = "1"
+    )]
+    pub cert_per_batch: u64,
+    /// Number of subnets to use for certificate generation. For every certificate sunbet id will be picked randomly.
```
Comment (Freyskeyd) [inline crates/topos/src/components/network/commands/spam.rs:36]:
```suggestion
    /// Number of subnets to use for certificate generation. For every certificate subnet id will be picked randomly.
```

File: crates/topos-certificate-spammer/README.md [hunk 3]
```diff
@@ -2,58 +2,54 @@
...
 
-The Certificates are dispatched to one node (using `--target-node <tce endpoint http address>`) or multiple nodes listed in the file given as
-argument `--target-node-list cert-spammer/example_target_nodes.json`.
+Source subnet id and list of target subnets are randomly assigned to every certificate.
+
+Argument `--nb-batches` is specified, program will send specified number of batches and then gracefully shut down connections and exit. Otherwise, it will continuously generate and send batches of certificates.
+
+The time delay in milliseconds between dispatching two certificates is set with `--batch-time-interval`.
+
+Certificates are signed with secp256k1 private key, and seed for generation of `nb-subnets` private keys is infuenced by `--local-key-seed`.
```
[the remark this answers is not in the document]
Comment (atanmarko) [reply crates/topos-certificate-spammer/README.md:14]: I am open to discussion on this topic

File: crates/topos-certificate-spammer/README.md
[the remark this answers is not in the document]
Comment (atanmarko) [reply crates/topos-certificate-spammer/README.md:1]: I agree, but first we should stabilize it later we can move it there when things are well known.

File: crates/topos/src/components/network/commands/spam.rs [hunk 4]
```diff
@@ -2,35 +2,68 @@ use clap::Args;
...
+    /// Multiple nodes could be specified as comma separated list
+    /// e.g. `--target-nodes=http://[::1]:1340,http://[::1]:1341`
+    #[clap(
+        long,
+        env = "TOPOS_NETWORK_SPAMMER_TARGET_NODES",
+        value_delimiter = ','
+    )]
+    pub target_nodes: Option<Vec<String>>,
+    /// Path to json file with list of target nodes as alternative to `--target-nodes`
+    #[clap(long, env = "TOPOS_NETWORK_SPAMMER_TARGET_NODES_PATH")]
```
[the remark this answers is not in the document]
Comment (atanmarko) [reply crates/topos/src/components/network/commands/spam.rs:20]: I'll open another ticket to synchronize arguments with TCE. Tools team is currently using this format

File: crates/topos/src/components/network/commands/spam.rs [hunk 5]
```diff
@@ -2,35 +2,68 @@ use clap::Args;
...
-    pub node_per_cert: usize,
-    /// Socket of the opentelemetry agent endpoint
+    /// List of generated certificate target subnets. No target subnets by default.
+    /// For example `--target-subnets=0x3bc19e36ff1673910575b6727a974a9abd80c9a875d41ab3e2648dbfb9e4b518,0xa00d60b2b408c2a14c5d70cdd2c205db8985ef737a7e55ad20ea32cc9e7c417c`
+    #[arg(
+        long,
+        env = "TOPOS_NETWORK_SPAMMER_TARGET_SUBNETS",
+        value_delimiter = ','
+    )]
+    pub target_subnets: Option<Vec<String>>,
```
[the remark this answers is not in the document]
Comment (atanmarko) [reply crates/topos/src/components/network/commands/spam.rs:61]: I'll open another ticket to synchronize arguments with TCE.

File: crates/topos-certificate-spammer/README.md [hunk 6]
```diff
@@ -2,58 +2,54 @@
...
-The Topos Certificate Spammer generates multiple Certificate per batch defined by `--cert-per-batch <number>`.
-The time delay in seconds between batches is controlled by `--batch-time-interval`.
+The Topos Certificate Spammer generates test certificate chain and sends them to one or more target nodes, specified with parameter `--target-nodes`, e.g. `--target-nodes=http://[::1]:1340,http://[::1]:1341`.
+Every `--batch-time-interval` one batch of `--cert-per-batch` certificates is generated.
 
-The Certificates are dispatched to one node (using `--target-node <tce endpoint http address>`) or multiple nodes listed in the file given as
-argument `--target-node-list cert-spammer/example_target_nodes.json`.
+Source subnet id and list of target subnets are randomly assigned to every certificate.
+
+Argument `--nb-batches` is specified, program will send specified number of batches and then gracefully shut down connections and exit. Otherwise, it will continuously generate and send batches of certificates.
```
[the remark this answers is not in the document]
Comment (atanmarko) [reply crates/topos-certificate-spammer/README.md:10]: I'll prefer to leave `--nb-batches`, as we also have `--nb-subnets` and just `--subnets` would be a bit confusing

File: crates/topos-certificate-spammer/README.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (Freyskeyd) [reply crates/topos-certificate-spammer/README.md:14]: let's create a spike. cc @hadjiszs

**Review by Freyskeyd — APPROVED** [review]

**Pull request merged**